Complete sprint 9

Evo-Eclipse opened the pull request:
### Overview of Changes

This pull request implements parts of key functional requirements for the Image Feed application:

User Profile Screen

- Created a new `ProfileViewController` in Storyboard
- Added core UI elements:
    - UIImageView for profile avatar
    - Three UILabels to display user information
    - UIButton for logout functionality
- Configured correct Constraints for all elements
- Integrated resources from Figma:
    - Profile image
    - Logout button icon

Single Image Screen

- Implemented `SingleImageViewController`
- Added ScrollView with Constraints set to zero on all sides
- Configured image display:
    - Stretching to screen boundaries
    - Centering the image
- Created outlet for ScrollView

Sharing Functionality

- Added "Share" button on SingleImageViewController
- Implemented `didTapShareButton` method
- Set up image sharing functionality
- Positioned button above ScrollView

Implementation Details

- Updated `Main.storyboard` with new scenes
- Added necessary screen transitions
- Adhered to UI/UX requirements from the design mockup

Evo-Eclipse committed:
Add ProfileViewController

- Create class and update Main.storyboard
- Set UI elements and constraints
- Connect outlets and actions

Evo-Eclipse committed:
Add UITabBarController

- Create UITabBarController and update storyboard
- Configure tab bar items
- Set UITabBarController as the initial view controller
- Adjust appearance to match the design

Evo-Eclipse committed:
Add UINavigationController

- Embed UINavigationController and update storyboard
- Adjust appearance to match the design

Evo-Eclipse committed:
Add SingleImageViewController

- Create SingleImageViewController and update storyboard
- Enable transition on cell tap

Evo-Eclipse committed:
Revise navigation approach

- Remove UINavigationController from Storyboard
- Replace push segue with modal presentation
- Add backward button assets
- Add back button for modal dismissal

Evo-Eclipse committed:
Add image zooming and sharing

- Wrapped Image View in Scroll View to enable image scaling functionality
- Implemented automatic image scaling and centering logic
- Added Share Button assets and sharing functionality

Evo-Eclipse committed:
Refactored profile screen to code

- Migrated Profile screen from storyboard to programmatic UI in ProfileViewController.swift
- Removed storyboard content for Profile screen
- Renamed image assets for easier access

Evo-Eclipse committed:
Add full-screen opening to SingleImageView

So1ovyov commented on ImageFeed/Presentation/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    
+    // MARK: - Private Properties
+    
+    private let profileImage = UIImageView()
```
Предлагаю попробовать настраивать `UI` элементы не в отдельных методах, а сразу при объявлении, например вот так:
```
    private let avatarImageView: UIImageView = {
        let view = UIImageView()
        view.image = UIImage(named: "icon.user")
        view.contentMode = .scaleAspectFit
        return view
    }()

   /// и т.д. для других UI элементов
```
Таким образом метод `setupProfileImage` можно вообще удалить, как и другие методы

So1ovyov commented on ImageFeed/Presentation/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    private func setupProfileImage() {
+        profileImage.image = UIImage(named: "icon.user")
+        profileImage.contentMode = .scaleAspectFit
+        profileImage.translatesAutoresizingMaskIntoConstraints = false
```
Обрати внимание на инструмент `forEach` при верстке и настройке элементов, он поможет тебе сократить повторяющийся код
Вот пример:
```
        [view1, 
         view2,
         view3].forEach {
            $0.translatesAutoresizingMaskIntoConstraints = false
             view.addSubview($0)
        }
```
Таким образом все строки `translatesAutoresizingMaskIntoConstraints` и `addSubview` можно удалить и очень сильно сократить код

So1ovyov commented on ImageFeed/Presentation/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,102 @@
+        
+        view.addSubview(profileImage)
+        
+        NSLayoutConstraint.activate([
```
Здесь можно с помощью инструмента `NSLayoutConstraints.activate` активировать сразу все констрейнты, а не каждый отдельно:
```
        private func setupСonstraints(){
            NSLayoutConstraint.activate([
                profileImage.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 32),
                profileImage.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
                profileImage.widthAnchor.constraint(equalToConstant: 70),
                profileImage.heightAnchor.constraint(equalToConstant: 70)

                profileName.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
                profileName.topAnchor.constraint(equalTo: avatarImageView.bottomAnchor, constant: 8),
                profileName.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -16),

                /// и т.д. с другими элементами продолжить в рамках этого массива
            ])
        }
```

So1ovyov commented on ImageFeed/Presentation/SingleImage/SingleImageViewController.swift:
```diff
@@ -0,0 +1,118 @@
+    
+    @IBAction private func didTapBackButton() {
+        dismiss(animated: true, completion: nil)
+    }
```
Предлагаю методы и структуры разделять между собой одной пустой строкой, чтобы было легче воспринимать код и где начинается/заканчивается каждая сущность

So1ovyov commented:
@Evo-Eclipse отличная работа, успехов в следующем спринте!

Evo-Eclipse committed:
Refactor ProfileViewController UI implementation

- Replace individual setup methods with property initializers
- Configure UI elements during declaration using closure pattern
- Consolidate all layout constraints into a single NSLayoutConstraint.activate call
- Improve code organization and readability

Evo-Eclipse commented:
Thank you for review!
